src/backend/distributed/commands/parameter.c

```diff
@@ -0,0 +1,178 @@
+#include "postgres.h"
+
+#include "pg_version_constants.h"
+#if PG_VERSION_NUM >= PG_VERSION_15
+#include "access/genam.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_parameter_acl.h"
+#include "commands/defrem.h"
+#include "utils/acl.h"
+#include "utils/builtins.h"
+#include "utils/syscache.h"
+
+#include "distributed/commands.h"
+#include "distributed/deparser.h"
+#include "distributed/grant_utils.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_sync.h"
+
+static List * GenerateGrantOnParameterFromAclItem(char *parameterName, AclItem *aclItem);
+static bool HasAclGrantOption(AclItem *aclItem, AclMode aclMode);
+static void CheckPermissionsAndGrants(AclItem *aclItem, AclMode modes[], int numModes);
+static void CheckAndAppendQuery(List **queries, AclItem *aclItem, Oid granteeOid,
+								char *parameterName, AclMode mode, char *modeStr);
+
+
+List *
+PostprocessGrantParameterStmt(Node *node, const char *queryString)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsurePropagationToCoordinator();
+
+	char *command = DeparseTreeNode(node);
+
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) command,
+								ENABLE_DDL_PROPAGATION);
+
+	return NontransactionalNodeDDLTaskList(REMOTE_NODES, commands);
+}
+
+
+/*
+ * GenerateGrantOnParameterFromAclItem generates a query string for replicating a users permissions
+ * on a database.
+ */
+static List *
+GenerateGrantOnParameterFromAclItem(char *parameterName, AclItem *aclItem)
+{
+	/*
+	 * seems unlikely but we check if there is a grant option in the list without the actual permission
+	 */
+	CheckPermissionsAndGrants(aclItem, (AclMode[]) { ACL_SET, ACL_ALTER_SYSTEM }, 2);
+	Oid granteeOid = aclItem->ai_grantee;
+	List *queries = NIL;
+
+	queries = lappend(queries, GenerateSetRoleQuery(aclItem->ai_grantor));
+
+	CheckAndAppendQuery(&queries, aclItem, granteeOid, parameterName, ACL_SET, "SET");
+	CheckAndAppendQuery(&queries, aclItem, granteeOid, parameterName, ACL_ALTER_SYSTEM,
+						"ALTER SYSTEM");
+
+	queries = lappend(queries, "RESET ROLE");
+
+	return queries;
+}
+
+
+/*
+ * CheckAndAppendQuery checks if the aclItem has the given mode and if it has, it appends the
+ * corresponding query to the queries list.
+ */
+static void
+CheckAndAppendQuery(List **queries, AclItem *aclItem, Oid granteeOid, char *parameterName,
+					AclMode mode, char *modeStr)
+{
+	AclResult aclresult = pg_parameter_aclcheck(parameterName, granteeOid, mode);
+	if (aclresult == ACLCHECK_OK)
+	{
+		char *query = DeparseTreeNode((Node *) GenerateGrantStmtForRightsWithObjectName(
+										  OBJECT_PARAMETER_ACL, granteeOid, parameterName,
+										  modeStr,
+										  HasAclGrantOption(aclItem, mode)));
+
+		/* remove the semicolon at the end of the query since it is already */
+		/* appended in metadata_sync phase */
+		query[strlen(query) - 1] = '\0';
+
+		*queries = lappend(*queries, query);
+	}
```

Comment:
Especially cannot quite follow what's happening here, could you simplify GenerateGrantOnParameterFromAclItem & CheckAndAppendQuery a bit?

Reply:
I simplified and added comments renamed functions and added some additional functions.